opentracing-impl-java8/src/main/java/io/opentracing/propagation/AbstractTracer.java

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.propagation;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import io.opentracing.Tracer.SpanBuilder;
+
+abstract class AbstractTracer implements Tracer {
+
+    static final boolean BAGGAGE_ENABLED = !Boolean.getBoolean("opentracing.propagation.dropBaggage");
+
+    private final PropagationRegistry registry = new PropagationRegistry();
+
+    protected AbstractTracer() {
+        registry.register(Map.class, new SplitTextFormatInjectorImpl());
+        registry.register(Map.class, new SplitTextFormatExtractorImpl());
+    }
+
+    @Override
+    public SpanBuilder buildSpan(String operationName){
+        return new SpanBuilderImpl().withOperationName(operationName);
+    }
+
+    @Override
+    public <T> void inject(Span span, T carrier) {
+        registry.getInjector((Class<T>)carrier.getClass()).injectSpan(span, carrier);
+    }
+
+    @Override
+    public <T> SpanBuilder join(T carrier) {
+        return registry.getExtractor((Class<T>)carrier.getClass()).joinTrace(carrier);
+    }
+
+    public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+        return registry.register(carrierType, injector);
+    }
+
+    public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+        return registry.register(carrierType, extractor);
+    }
+
+    private static class PropagationRegistry {
+
+        private final ConcurrentMap<Class, Injector> injectors = new ConcurrentHashMap<>();
+        private final ConcurrentMap<Class, Extractor> extractors = new ConcurrentHashMap<>();
```

Comment:
In fact, I would prefer to have a Builder pattern for tracer, so that the only way to register serializers is via builder, and in tracer these maps become immutable.